core/src/main/java/io/questdb/cairo/wal/SymbolMap.java

```diff
@@ -0,0 +1,213 @@
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2025 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.cairo.wal;
+
+import io.questdb.std.AbstractOffsetCharSequenceHashSet;
+import io.questdb.std.Chars;
+import io.questdb.std.MemoryTag;
+import io.questdb.std.Numbers;
+import io.questdb.std.Unsafe;
+import io.questdb.std.str.DirectString;
+import org.jetbrains.annotations.NotNull;
+
+import java.io.Closeable;
+import java.util.Arrays;
+
+public class SymbolMap extends AbstractOffsetCharSequenceHashSet implements Closeable {
+    public static final int NO_ENTRY_VALUE = -1;
+    private final int noEntryValue;
+    private final DirectString sview = new DirectString();
+    private long address;
+    private long charsCapacity;
+    private int currentOffset = 0;
+    private int[] values;
+
+    public SymbolMap() {
+        this(8);
+    }
+
+    public SymbolMap(int initialCapacity) {
+        this(initialCapacity, 0.4, NO_ENTRY_VALUE);
+    }
+
+    public SymbolMap(int initialCapacity, double loadFactor, int noEntryValue) {
+        this(initialCapacity, loadFactor, noEntryValue, 32);
+    }
+
+    public SymbolMap(int initialCapacity, double loadFactor, int noEntryValue, int avgKeySize) {
+        super(initialCapacity, loadFactor);
+        this.noEntryValue = noEntryValue;
+        this.charsCapacity = Numbers.ceilPow2(initialCapacity * (avgKeySize + 8));
+        this.address = Unsafe.malloc(this.charsCapacity, MemoryTag.NATIVE_DEFAULT);
+        values = new int[offsets.length];
+    }
+
+    @Override
+    public final void clear() {
+        super.clear();
+        Arrays.fill(values, noEntryValue);
+        this.currentOffset = 0;
+    }
+
+    @Override
+    public void close() {
+        if (this.address != 0) {
+            Unsafe.free(address, this.charsCapacity, MemoryTag.NATIVE_DEFAULT);
+            this.address = 0;
+            this.charsCapacity = 0;
+        }
+    }
+
+    public int get(@NotNull CharSequence key) {
+        return valueAt(keyIndex(key));
+    }
+
+    public CharSequence get(int offset) {
+        final long lo = address + ((long) offset << 2);
+        final int len = Unsafe.getUnsafe().getInt(lo + 4);
```

Comment:
Please add and assert that len does not point beyond the capacity.
Also, please assert that it's non-negative.